File: Practice/Memoization/HowSum.py
def howSum(targetSum, arr, memo):
    # print(memo)
    if targetSum in memo:
        print(memo)
        return memo[str(targetSum)]

    if targetSum == 0:
        return []

    if targetSum < 0:
        return None

    print(memo)
    for num in arr:
        remainder = targetSum - num
        remainderResult = howSum(remainder, arr, memo)
        if(remainderResult != None):
            ans = remainderResult.copy()
            ans.append(num)
            print(memo)
            memo[str(targetSum)] = ans
            return memo[str(targetSum)]

    memo[str(targetSum)] = None
    return None


# print(howSum(7, [2, 3], {}))
# print(howSum(7, [5, 4, 3, 7], {}))
# print(howSum(7, [2, 4], {}))
# print(howSum(8, [2, 3, 5], {}))
print(howSum(300, [7, 14], {}))

File: Practice/Memoization/HowSum.js
const howSum = (targetSum, arr, memo) => {
    // console.log(memo);
    if(targetSum in memo) return memo[targetSum];
    
    if(targetSum === 0){
        return [];
    }
    if(targetSum < 0){
        return null;
    }
    

    for (let num in arr){
        const remainder = targetSum - num;
        const remainderResult = howSum(remainder, arr, memo);
            if(remainderResult !== null){
                console.log(remainderResult);
                memo[targetSum] = [ ...remainderResult, num ];
                return memo[targetSum];
            }
    }

    memo[targetSum] = null;
    return null;
}
// Some issue

console.log(howSum(7, [2, 3], {}));
// console.log(howSum(7, [5, 3, 4, 7]));
// console.log(howSum(7, [2, 4]));
// console.log(howSum(8, [2, 3, 5]));
// console.log(howSum(300, [7, 14]));
